skopt/space/space.py

```diff
@@ -721,3 +724,108 @@ def distance(self, point_a, point_b):
             distance += dim.distance(a, b)
 
         return distance
+
+    def get_dimension(self, id):
+        """Get the Dimension-object in the search-space using either
+        its name or index as the lookup-id.
+
+        Parameters
+        ----------
+        * `id` str or int:
+            If `id` is a string then it is the name of a dimension.
+            If `id` is an integer then it is the index of a dimension.
+
+        Returns
+        -------
+        * (`dimension`, `index`, `name`) [tuple(Dimension, int, str)]:
```

Comment:
we should return just the dimension

Reply:
Well, the reason I do it like this is to try and provide efficient support for the existing plotting functions. I agree that it is an ugly function. I'll see if I can fix the plotting functions to make this better.

Reply:
The index is now set inside each `Dimension`-instance by `Space.__init__` as we need it for e.g. plotting.

Reply:
I don't think a `Dimension` inside a space should know what its index is. If you need the index when iterating over the dimensions in a space you should use:
```python
for i, dim in enumerate(space):
  print(i, dim)
```

Reply:
The reason I originally made the above function return both the index and the `Dimension`-objects was precisely because I did not want to modify your code. But you need the index for various things when only using a subset of the dimensions and you cannot loop in the way you propose. So an efficient way of doing it was the above, but the API was ugly as I agreed.

What I have done now is to modify both `Dimension` and `Space` to give each `Dimension`-instance an index once they are added to a `Space`. This way we have easy access to it when we need it e.g. for plotting a sub-set of the dimensions, and for having default names such as X_i for unnamed dimensions. I think it is a reasonable solution.

The only problem with this would be if you later want to change the composition of dimensions in a search-space. But I don't see any need for doing that, or why that should even be supported by the API.

Are there any other problems with having an index inside the `Dimension`-object?